### DISCUSSION TITLE
Parsing bits from bytes

### DISCUSSION THREAD

**Discussion by TomasAdvance** [frangoteam/FUXA#295] [General]
Hi Fuxa team,
Exist any possibilities to parsing specific bits from bytes? For example, if I have coded some information in the byte (bit 0 think A, bit 1 means think B etc).
Thanks for answer
T

**Comment by unocelli** [comment]
Hi, could you explain the reason, optimise resources on the plc? I know that for alarms it is used (1 Bit for alarm).
can you give me an example of how you would like to use it on the gui?
Thanks

**Reply by TomasAdvance** [reply]
Hi,
For example, in some our system se have coded alarm and status od the devices in separate bits located in holding registers not in coil registers.
In that case we se some address in holding register where individual bits represent Sam status like - bit 0 = breaker On, bit 1 breaker Off, bit 2 = alarm 1 etc.
Is it more clear right now?
I am looking for read individual bits or parsing individual bits in holding registers with read and write capabilities. 
T

**Comment by ihor-horash** [comment]
Hi, parsing bits from int or byte is very common scenario in industry because a lot of PLC's libraries has written in that way when device status bits are packed into the one integer value, like @TomasAdvance mentioned. It will be very usefull feature to use existing PLC libraries with FUXA

**Reply by unocelli** [reply]
Hi,
I also see the functionality of masking bits to identify states or alarms, and it shouldn't be too complicated to add the feature, but it gets complicated if for example you want to use the first 5 bits to define the value of a valve.

**Reply by ihor-horash** [reply]
Thank you for reply! What do you mean the "value of valve"? Is it about valve status? For example we have integer tag which bit0 - valve closed, bit1 - valve opened and bit2 - valve fault etc and we define rules with this bits to fill valve with specific colors. So if both bit0 (closed) is true and bit2 (fault) is true, what color should be chosen to fill the valve? I think the solution is to use the last defined rule to fill the valve.

**Reply by unocelli** [reply]
I was thinking of value to define the opening in %. but could be the rpm of a motor. but I think I understand what you mean ;)

**Reply by ihor-horash** [reply]
ok:) Your project is awesome!

**Reply by unocelli** [reply]
:pray:

**Comment by unocelli** [comment]
I am trying to implement the feature, the idea is to match the bitmask to the tag and not to each different status (its easier).
what do you think about?
![image](https://user-images.githubusercontent.com/48684550/161443085-35e032be-869d-4a0c-a325-e1d6482d55a8.png)

![image](https://user-images.githubusercontent.com/48684550/161443121-3907039a-2aa5-41c7-a321-d0ee0af74b94.png)

**Reply by ihor-horash** [reply]
Looks great!

**Comment by robsori** [comment]
The bit mask looks good. I would suggest the following:
1. The bit mask should start at 0 (0-31).
2. In the case of a byte tag, check boxes 8 to 31 should be blocked (readonly)
3. In the case of a word tag, only check boxes 16 to 31 should be blocked (readonly)
4. In case of a dword type tag, the bit mask (0-31) should remain unchanged

**Reply by unocelli** [reply]
Thanks for the suggestions,
1. I have some reservations, because we think as programmer
2-4 TODO

**Comment by TomasAdvance** [comment]
Hi it sounds very good, I agree.

Tomas

Tomas Pastrnak

Dne 03. 04. 22 v 20:43 Umberto Nocelli napsal(a):
> am trying to implement the feature, the idea is to match the bitmask 
> to the tag and not to each different status (its easier)

**Comment by IonuDragomir1** [comment]
Hi Unocelli,

Thank you for the effort of implementing bitmask!
I've tested it and I'd like to understand how it's working.
The test I've run are the following:
![image](https://user-images.githubusercontent.com/58516894/162471436-d283ce76-b44a-466d-af99-67f09f1cc3db.png)
I have encountered some problems:
1. The bitmask doesn't start from 0. As you can see, in PLC programming, the offset starts from 0, not 1.
2. It's a bit confusing how is made the bit relation between bitmask and bit selected (from 1 to 32). For example, if a user selects bit 9 its mask is 100. Instead of 100 should be displayed the value of selected bit, meaning 9.
3. If I change the color of an object, it would be more convenient to choose the value of the selected bit (1 or 0) and not the value of the word. It's not so intuitive what value should be set. 
4. It would be great if this bitmask feature could be added also for alarms.

**Reply by unocelli** [reply]
Hi, I'm sorry for the delay. Thanks for the report. 
1 I think you are right.
2 - 3 do you think it is used by only setting a single bit? if yes, then your suggestion is better.
4 of course

**Comment by yoabau** [comment]
Hello everyone,

Thank you @IonuDragomir1  for your comment. It really helped me understanding the bitmask feature. The feature actually works as query of bit as defined [her](https://en.wikipedia.org/wiki/Mask_(computing)#Querying_the_status_of_a_bit). After selecting the required bit to monitor, the shape can return a color based on status (0=False or True=1) of one of the bits. Reusing your example I managed to get this result:
![2023-10-26_13h14_50](https://github.com/frangoteam/FUXA/assets/65852298/a22928ac-36a6-4035-ae27-919f4d38a2c6)

@unocelli As you can see "shape_10" returns green color since bit 0 and 1 are both True.
But I was expecting the function to return decimal value based on bit 0 and 1 regardless the status of bit 2 to 31 such as:
bit 0 | bit 1 | bit 2-31| value(dec) |
|-----|-----|-----|-----|
| 0 | 0 | x | 0 |
| 1 | 0 | x | 1 |
| 0 | 1 | x | 2 |
| 1 | 1 | x | 3 |

But it doesn't, and instead it seems to only check whether one of the bit has value 1=True?

**Comment by unocelli** [comment]
Hi, 
I think there are different interpretations of bitmask management and in any case it is not possible to satisfy everyone.
The bitmask is intended to be associated with the value of the the bit (1/0) and not with the value masked with the bitmask.

**Reply by yoabau** [reply]
Hi,
Thank you for your answer. I raised an issue #954 to further developed this topic. It is not a question of interpretation but compliance with industry standards. However FUXA is free to develop its tool :)
